% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ATAC.R
\name{ATAC_Add_meta_data}
\alias{ATAC_Add_meta_data}
\title{Add Meta Data Information onto a SeuratObject}
\usage{
ATAC_Add_meta_data(sg_lib, mtx, fragments, replicate = 1, cal.FRiP = TRUE)
}
\arguments{
\item{sg_lib}{Data frame or directory to a txt file containing 3 columns: cell, barcode, gene. If sgRNA information stored in a matrix-like format or input data frame only has sgRNA frequency of each cell, use \code{\link[SCREE]{sgRNAassign}} to assign sgRNA to each cell.}

\item{mtx}{SeuratObject or directory to rds file of SeuratObject, with cell in columns and features in rows.}

\item{fragments}{Directory of fragments file.}

\item{replicate}{Directory of a txt file or a vector only containg the replicate information of each cell, in the same order of cells in SeuratObject. If no replicate information, we will consider that all cells are from the same replicate, and this parameter will be set as 1. Default is 1.}

\item{cal.FRiP}{Logical, calculate FRiP or not. Default is \code{TRUE}.}
}
\description{
Add sgRNA information, fractions of reads in peaks (FRiP) and replicate information onto a SeuratObject for scATAC-seq based input.
}
