#' Generate HTML Config 
#'
#' Generate config for HTML output, based on previous output of SCREE.
#' 
#' @param mtx SeuratObject or directory to rds file of SeuratObject, with cell in columns and features in rows. 
#' @param mtx_QC SeuratObject or directory to rds file after QC, with cell in columns and features in rows.
#' @param sg_lib Data frame or directory to a txt file containing 3 columns: cell, barcode, gene. If sgRNA information stored in a matrix-like format or input data frame only has sgRNA frequency of each cell, use \code{\link[SCREE]{sgRNAassign}} to assign sgRNA to each cell.
#' @param score Data frame or directory of score from \code{improved_scmageck_lr}.
#' @param pval Data frame or directory of p_value from \code{improved_scmageck_lr}.
#' @param project Project name.
#' @param prefix Path to the results generated by SCREE. Default is current directory.
#' @param label The prefix label of previous output file. Notably, there needs a separator between default file names and the label, so label would be better to be like "label_". Default is "".
#' @param species Only support "Hs" and "Mm". Default is "Hs".
#' @param version Version of the reference genome(Ensembl). Default is "v75".
#' @param type Data type, one of "RNA", "ATAC", "enhancer". Default is "ATAC".
#' @param NTC The name of negative controls. Default is "NTC".
#' @param article Which article the dataset from (URL).
#' @param data Data source.
#' @param article_name Name of the article.
#' @param data_name Name of the data.
#' @param gene_type Type of gene name, selected from one of c("Symbol", "Ensembl"). Default is "Symbol".
#' @param score_cut Score cutoff of \code{improved_scmageck_lr} results. Default is 0.2.
#' @param pval_cut P-value cutoff of \code{improved_scmageck_lr} results. Default is 0.05.
#' @param DA DA peak config generated by SCREE. Default is \code{NULL}.
#' @param cicero Cicero config generated by SCREE. Default is \code{NULL}.
#' @param enhancer Enhancer config generated by SCREE. Default is \code{NULL}.
#'
#' @importFrom utils read.table write.table
#' @export

config_generation <- function(mtx, mtx_QC, sg_lib, score, pval, project, prefix = ".", label = "", species = "Hs", version = "v75", type = "ATAC", NTC = "NTC", article = "", data = "", article_name = "", data_name = "", gene_type = "Symbol", score_cut = 0.2, pval_cut = 0.05, DA = NULL, cicero = NULL, enhancer = NULL, base64 = FALSE) {
    
    if (is.character(mtx)) {
        message(paste("Reading RDS file:", mtx))
        mtx <- readRDS(mtx)
    }
    
    if (is.character(mtx_QC)) {
        message(paste("Reading RDS file:", mtx_QC))
        mtx_QC <- readRDS(mtx_QC)
    }
    
    if (is.character(sg_lib)) {
        message(paste("Reading sgRNA lib file:", sg_lib))
        sg_lib <- read.table(sg_lib, header = T)
    }
    
    PerturbGene <- length(unique(sg_lib$gene)) - length(NTC)
    sg_num <- length(unique(sg_lib$barcode))
    NTC_num <- length(unique(subset(sg_lib, gene %in% NTC)))
    
    #generate config of html
    
    if (species == "Hs") {
        if (version == "v75") {
            gene_anno <- "EnsDb.Hsapiens.v75(hg19)"
        } else if(version == "v79") {
            gene_anno <- "EnsDb.Hsapiens.v79(hg38)"
        } else if(version == "v86") {
            gene_anno <- "EnsDb.Hsapiens.v86(hg38)"
        }
    } else if(species == "Mm") {
        if (version == "v75") {
            gene_anno <- "EnsDb.Mmusculus.v75(mm9)"
        } else if(version == "v79") {
            gene_anno <- "EnsDb.Mmusculus.v79(mm10)"
        }
    }

    if (gene_type == "Symbol") {
        gene_label <- "Gene Symbol"
    } else if (gene_type == "Ensembl") {
        gene_label <- "Ensembl ID"
    }

    if (article != "") {
        if (article_name == "") {
            article_name <- article
        }
        article <- paste('<a href=\"', article, '\" target="view_window">', article_name, '</a>', sep = "")
        #article <- paste0("'", article, "'")
    }
    
    if (data != "") {
        if (data_name == "") {
            data_name <- data
        }
        data <- paste('<a href=\"', data, '\" target="view_window">', data_name, '</a>', sep = "")
        #data <- paste0("'", data, "'")
    }

    if (type == "ATAC") {

        config <- c(article, data, PerturbGene, sg_num, NTC_num, prefix, "single-cell CRISPR ATAC-seq screens", 
                    project, gene_anno, gene_label, score_cut, pval_cut)
        names(config) <- c("ArticlePath", "GEO", "PerturbGene", "sgNumbers", "NTC", 
                           "ResultsPath", "DataType", "Project", "ReferenceVersion",
                           "GeneVersion", "Score_cut", "pval_cut")
        
    } else if (type == "enhancer") {
        
        config <- c(article, data, PerturbGene, sg_num, NTC_num, prefix, "single-cell CRISPR RNA-seq screens", 
                    project, gene_anno, gene_label, score_cut, pval_cut)
        names(config) <- c("ArticlePath", "GEO", "PerturbGene", "sgNumbers", "NTC", 
                           "ResultsPath", "DataType", "Project", "ReferenceVersion",
                           "GeneVersion", "Score_cut", "pval_cut")
    } else if (type == "RNA") {
        config <- c(article, data, PerturbGene, sg_num, NTC_num, prefix, "single-cell CRISPR RNA-seq screens", 
                    project, gene_anno, gene_label, score_cut, pval_cut)
        names(config) <- c("ArticlePath", "GEO", "PerturbGene", "sgNumbers", "NTC", 
                           "ResultsPath", "DataType", "Project", "ReferenceVersion",
                           "GeneVersion", "Score_cut", "pval_cut") 
    }

    config <- paste(names(config), config, collapse = "\' , \'", sep = "\' : \'")
    config <- paste("\"Config\" : {\'", config, "\'}", sep = "")
        
        
    #generate single-cell quality of html
    
    if (type == "ATAC") {
        vln <- file.path("results", "ATAC_quality", "img", paste(label, "raw_matrix_quality_vlnplot.png", sep = ""))
        frag <- file.path("results", "ATAC_quality", "img", paste(label, "FragmentsSize.png", sep = ""))
        umap1 <- file.path("results", "ATAC_quality", "UMAP", "img", paste(label, "umap_seurat_clusters.png", sep = ""))
        umap2 <- file.path("results", "ATAC_quality", "UMAP", "img", paste(label, "umap_perturbations.png", sep = ""))
        
        if (base64 == TRUE) {
            vln <- knitr::image_uri(file.path(prefix, vln))
            frag <- knitr::image_uri(file.path(prefix, frag))
            umap1 <- knitr::image_uri(file.path(prefix, umap1))
            umap2 <- knitr::image_uri(file.path(prefix, umap2))
        }
        
        sg_figure <- c(vln, frag, umap1, umap2)
        names(sg_figure) <- c("vlnplot", "fragments", "UMAP1", "UMAP2")
    } else {
        vln <- file.path("results", "quality", "img", paste(label, "raw_matrix_quality_vlnplot.png", sep = ""))
        umap1 <- file.path("results", "quality", "UMAP", "img", paste(label, "umap_seurat_clusters.png", sep = ""))
        umap2 <- file.path("results", "quality", "UMAP", "img", paste(label, "umap_perturbations.png", sep = ""))
        
        if (base64 == TRUE) {
            vln <- knitr::image_uri(file.path(prefix, vln))
            umap1 <- knitr::image_uri(file.path(prefix, umap1))
            umap2 <- knitr::image_uri(file.path(prefix, umap2))
        }
        
        sg_figure <- c(vln, umap1, umap2)
        names(sg_figure) <- c("vlnplot", "UMAP1", "UMAP2")
    }


    #get information of quality plot
    
    sg_figure <- paste(names(sg_figure), sg_figure, collapse = "\" , \"", sep = "\" : \"")
    sg_figure <- paste("\"figure\" : {\"", sg_figure, "\"}", sep = "")

    #get information of cells and genes
    
    info <- c(project, ncol(mtx), nrow(mtx), ncol(mtx_QC), nrow(mtx_QC))
    names(info) <- c("SampleName", "RawCell", "RawGene", "QCCell", "QCGene")
    info <- paste(names(info), info, collapse = "\" , \"", sep = "\" : \"")
    info <- paste("\"quality\" : {\"", paste(info, sg_figure, sep = "\" , "), "}", sep = "")

    #generate sgRNA information of html
        
    img_prefix <- file.path("results", "sgRNA", "img", "sgRNA_quality_of_gene")
    sg_right <- c(file.path("results", "sgRNA", "img", paste(label, "cell_numbers.png", sep = "")))
    if (base64 == TRUE) {
        sg_right <- knitr::image_uri(file.path(prefix, sg_right))
    }
    sg_lib_filtered <- subset(sg_lib, cell %in% intersect(sg_lib$cell, colnames(mtx)))
    for (genes in sort(unique(sg_lib_filtered$gene))) {
        sg_right_genes <- file.path(img_prefix, paste(genes, ".png", sep = ""))
        if (base64 == TRUE) {
            sg_right_genes <- knitr::image_uri(file.path(prefix, sg_right_genes))
        }
        sg_right <- c(sg_right, sg_right_genes)
    }
    names(sg_right) <- c("Top10", sort(unique(sg_lib_filtered$gene)))
    sg_right <- paste(names(sg_right), sg_right, collapse = "\" , \"", sep = "\" : \"")
    sg_right <- paste("\"right\" : {\"", sg_right, "\"}", sep = "")
    sg_left_path <- file.path("results", "sgRNA", "img", paste(label, "sgRNA_numbers.png", sep = ""))
    if (base64 == TRUE) {
        sg_left_path <- knitr::image_uri(file.path(prefix, sg_left_path))
    }
    sg_left <- paste("left\" : \"", sg_left_path)
    sg <- paste("\"sgRNA\" : {\"", paste(sg_left, sg_right, sep = "\" , "), "}", sep = "")

    #generate Mixscape information of html
    
    if (file.exists(file.path(prefix, "results", "perturbation_efficiency", "mixscape", "img", 
                              paste(label, "mixscape_after.png", sep = "")))) {
        cluster <- c(file.path("results", "perturbation_efficiency", "mixscape", "img",
                               paste(label, "umap_seurat_clusters.png", sep = "")), 
                     file.path("results", "perturbation_efficiency", "mixscape", "img",
                                paste(label, "umap_perturbations.png", sep = "")))
        if (base64 == TRUE) {
            cluster <- c(knitr::image_uri(file.path(prefix, cluster[1])), knitr::image_uri(file.path(prefix, cluster[2])))
        }
        
        names(cluster) <- c("UMAP1", "UMAP2")
        cluster <- paste(names(cluster), cluster, collapse = "\" , \"", sep = "\" : \"")
        cluster <- paste("\"clustering\" : {\"", cluster, "\"}", sep = "")
        # umap <- c(file.path("results", "perturbation_efficiency", "mixscape", "img", 
        #                     paste(label, "mixscape_before.png", sep = "")), 
        #           file.path("results", "perturbation_efficiency", "mixscape", "img", 
        #                     paste(label, "mixscape_after.png", sep = "")),
        #           cluster)
        umap <- c(file.path("results", "perturbation_efficiency", "mixscape", "img", 
                            paste(label, "mixscape_before.png", sep = "")), 
                  file.path("results", "perturbation_efficiency", "mixscape", "img", 
                            paste(label, "mixscape_after.png", sep = "")))
        
        mixscape_heatmap <- c(file.path("results", "perturbation_efficiency", "img", 
                                        paste(label, "before_perturb_ratio.png", sep = "")),
                              file.path("results", "perturbation_efficiency", "img", 
                                        paste(label, "after_perturb_ratio.png", sep = "")))
        
        if (base64 == TRUE) {
            umap <- c(knitr::image_uri(file.path(prefix, umap[1])), knitr::image_uri(file.path(prefix, umap[2])))
            mixscape_heatmap <- c(knitr::image_uri(file.path(prefix, mixscape_heatmap[1])), 
                                  knitr::image_uri(file.path(prefix, mixscape_heatmap[2])))
        }
    } else {
        umap <- c(file.path("results", "perturbation_efficiency", "mixscape", "img", 
                            paste(label, "mixscape_before.png", sep = "")), "", "")
        mixscape_heatmap <- c("", "")
        if (base64 == TRUE) {
            umap <- c(knitr::image_uri(file.path(prefix, umap[1])), "", "")
        }
    }

    #names(umap) <- c("before", "after", "clustering")
    names(umap) <- c("before", "after")
    umap <- paste(names(umap), umap, collapse = "\" , \"", sep = "\" : \"")
    umap <- paste("\"UMAP\" : {\"", umap, "\"", " , ", cluster, "}", sep = "")
    names(mixscape_heatmap) <- c("before", "after")
    mixscape_heatmap <- paste(names(mixscape_heatmap), mixscape_heatmap, collapse = "\" , \"", sep = "\" : \"")
    mixscape_heatmap <- paste("\"heatmap\" : {\"", mixscape_heatmap, "\"}", sep = "")

    if (dir.exists(file.path(prefix, "results", "perturbation_efficiency", "mixscape", "img", "KO_percent", ""))) {
        l <- length(list.files(file.path(prefix, "results", "perturbation_efficiency", "mixscape", "img", "KO_percent", "")))
        ko <- c()
        for (i in 1 : l) {
            ko_path <- file.path("results", "perturbation_efficiency", "mixscape", "img", "KO_percent",
                                 paste(i, ".png", sep = ""))
            if (base64 == TRUE) {
                ko_path <- knitr::image_uri(file.path(prefix, ko_path))
            }
            ko <- c(ko, ko_path)
        }
        names(ko) <- seq(1 : l)
        ko <- paste(names(ko), ko, collapse = "\" , \"", sep = "\" : \"")
        ko <- paste("\"KO\" : {\"", ko, "\"}", sep = "")
    } else {
        ko <- "\"KO\" : \"\""
    }


    mixscape <- paste("\"Mixscape\" : {", paste(umap, mixscape_heatmap, ko, sep = " , "), "}", sep = "")

    #generate scMAGeCK information of html
    
    prefix_volcano <- file.path("results", "perturbation_efficiency", "volcano", "img")
    volcano <- c()
    for (genes in sort(colnames(score))) {
        volcano_genes <- file.path(prefix_volcano, paste(genes, ".png", sep = ""))
        if (base64 == TRUE) {
            volcano_genes <- knitr::image_uri(file.path(prefix, volcano_genes))
        }
        volcano <- c(volcano, volcano_genes)
    }
    names(volcano) <- sort(colnames(score))
    volcano <- paste(names(volcano), volcano, collapse = "\" , \"", sep = "\" : \"")
    volcano <- paste("\"volcano\" : {\"", volcano, "\"}", sep = "")

    if (file.exists(file.path(prefix, "results", "perturbation_efficiency", "img", 
                              paste(label, "correlation_heatmap.png", sep = "")))) {
        sc <- c(file.path("results", "perturbation_efficiency", "img", 
                          paste(label, "DE_gene_cutoff", score_cut, "_p", pval_cut, ".png", sep = "")),
                file.path("results", "perturbation_efficiency", "img", paste(label, "correlation_heatmap.png", sep = "")))
        if (base64 == TRUE) {
            sc <- c(knitr::image_uri(file.path(prefix, sc[1])), knitr::image_uri(file.path(prefix, sc[2])))
        }
    } else {
        sc <- c(file.path("results", "perturbation_efficiency", "img", 
                          paste(label, "DE_gene_cutoff", score_cut, "_p", pval_cut, ".png", sep = "")), 
                "")
        sc <- c(knitr::image_uri(file.path(prefix, sc[1])), "")
    }
    names(sc) <- c("barplot", "heatmap")
    sc <- paste(names(sc), sc, collapse = "\" , \"", sep = "\" : \"")
    scmageck <- paste("\"scMAGeCK\" : {\"", paste(sc, volcano, sep = "\" , "), "}", sep = "")
        
    #generate DE genes information of html
        
    results <- data.frame()
    GO <- c()
    prefix_GO <- file.path(prefix, "results", "potential_target_gene", "GO", "img")
    go <- list.files(file.path(prefix_GO))
    j <- 0
    k <- 0
    DE <- c()
    for (i in sort(colnames(score))) {
        a_scmageck <- data.frame(target = rownames(score), score = score[, i], p_val = pval[, i])
        a_scmageck <- subset(a_scmageck, abs(score) > score_cut & p_val < pval_cut)
        if (paste(i, ".png", sep = "") %in% go) {
            k <- k + 1
            GO_genes <- file.path("results", "potential_target_gene", "GO", "img", paste(i, ".png", sep = ""))
            
            if (base64 == TRUE) {
                GO_genes <- knitr::image_uri(file.path(prefix, GO_genes))
            }
            
            GO <- c(GO, GO_genes)
            names(GO)[k] <- i
        }
        if (nrow(a_scmageck) == 0) {
            next
        } else {
            j <- j + 1
            de <- data.frame(Perturbations = rep(i, nrow(a_scmageck)), 
                             target = a_scmageck$target, 
                             score = a_scmageck$score,
                             p_val = a_scmageck$p_val)
            results <- rbind(results, de)
            de <- paste("<tr><td>", de$Perturbations, 
                        "</td><td>", de$target, 
                        "</td><td>", round(de$score, 3), 
                        "</td><td>", de$p_val, 
                        "</td></tr>", sep = "")
            de <- paste(de, collapse = "", sep = "")
            DE <- c(DE, de)
            names(DE)[j] <- i
        }
    }
    if (nrow(results) != 0) {
        all_de <- paste("<tr><td>", results$Perturbations, 
                        "</td><td>", results$target, 
                        "</td><td>", round(results$score, 3), 
                        "</td><td>", results$p_val, 
                        "</td></tr>", sep = "")
        all_de <- paste(all_de, collapse = "", sep = "")
        DE <- c(all_de, DE)
        names(DE)[1] <- "All_Results"
        DE <- paste(names(DE), DE, collapse = "\" , \"", sep = "\" : \"")
        DE <- paste("\"table\" : {\"", DE, "\"}", sep = "")
    } else {
        DE <- paste("\"All_Results\" : ", "\"\"", sep = "")
        DE <- paste("\"table\" : {", DE, "}", sep = "")
    }
        

    if (length(GO) != 0) {
        GO <- paste(names(GO), GO, collapse = "\" , \"", sep = "\" : \"")
        GO <- paste("\"GO\" : {\"", GO, "\"}", sep = "")
    } else {
        GO <- "\"GO\" : \"\""
    }

    DEgenes <- paste("\"DEgenes\" : {", paste(DE, GO, sep = " , "), "}", sep = "")
        
    #generate enhancer results from output files
        
    if (is.null(enhancer)) {
        if (type == "enhancer") {
            en <- list.files(file.path(prefix, "results", "enhancer_function", "enhancer_gene_expression", "img", ""))
            all_enhancer <- c()
            for (perturb in en) {
                enhancer <- c()
                
                #generate enhancer directory of html
                
                enhancer_prefix <- file.path(file.path(prefix, "results", "enhancer_function", 
                                                       "enhancer_gene_expression", "img", ""), perturb)

                l <- length(list.files(file.path(enhancer_prefix)))
                for (i in 1 : l) {
                    enhancer_i <- paste(file.path("results", "enhancer_function", 
                                                  "enhancer_gene_expression", "img", perturb, i), ".png", sep = "")
                    
                    if (base64 == TRUE) {
                        enhancer_i <- knitr::image_uri(file.path(prefix, enhancer_i))
                    }
                    
                    enhancer <- c(enhancer, enhancer_i)
                    names(enhancer)[i] <- i
                }
                
                enhancer <- paste(names(enhancer), enhancer, collapse = "\" , \"", sep = "\" : \"")
                enhancer <- paste("\"", perturb, "\" : {\"", enhancer, "\"}", sep = "")
                all_enhancer <- c(all_enhancer, enhancer)
                
            }
            all_enhancer <- paste("", all_enhancer, collapse = ", ", sep = "")
            all_enhancer <- paste("\"enhancer\" : {", all_enhancer, "}", sep = "")
            enhancer <- all_enhancer
        }
    }
        
    #generate cicero results from output files
    
    if (is.null(cicero)) {
        if (type == "enhancer") {
            ci <- list.files(file.path(prefix, "results", "enhancer_function", "cicero", "img", ""))
            ci <- gsub(".png", "", ci)
            k <- 0
            cicero <- c()
            for(j in ci) {    
                
                #get html config
                
                k <- k + 1
                
                cicero_i <- file.path("results", "enhancer_function", "cicero", "img", paste(j, ".png", sep = ""))
                
                if (base64 == TRUE) {
                    cicero_i <- knitr::image_uri(file.path(prefix, cicero_i))
                }
                
                cicero <- c(cicero, cicero_i)
                names(cicero)[k] <- j
            }
            cicero <- paste(names(cicero), cicero, collapse = "\" , \"", sep = "\" : \"")
            cicero <- paste("\"cicero\" : {\"", cicero, "\"}", sep = "")
        }
    }
        
    #generate results information of html
        
    if (type == "enhancer") {
        results <- paste(info, sg, mixscape, scmageck, DEgenes, enhancer, cicero, sep = " , ")
    } else if (type == "ATAC") {
        results <- paste(info, sg, mixscape, scmageck, DEgenes, DA, cicero, sep = " , ")
    } else if (type == "RNA") {
        results <- paste(info, sg, mixscape, scmageck, DEgenes, sep = " , ")
    }
        
    #generate final config value
       
    results <- paste("\"Results\" : {", results, "}", sep = "")
    final <- paste(config, results, sep = " , ")
    final <- paste("\nvar SCREE_result = {", final, "};\n", sep = "")
        
    return(final)
        
}

#' Generate HTML
#'
#' Generate HTML based on the config from \code{\link[SCREE]{config_generation}}.
#'
#' @param html_dir Template HTML file directory.
#' @param config Config generated before.
#' @param replace Either the child number to return (by position), or the name of the child node to return. If there are multiple child nodes with the same name, the first will be returned
#' @param prefix Path to save the html. Default is current directory.
#' @param label The prefix label of the output file. Notably, there needs a separator between default file names and the label, so label would be better to be like "label_". Default is "".
#'
#' @import xml2
#' @export 


html_output <- function(html_dir, config, replace = 3, prefix = ".", label = "") {
    html <- xml2::read_html(html_dir)
    xml2::xml_set_text(xml2::xml_child(html, replace), config)
    xml2::write_html(html, file.path(prefix, paste(label, "summary.html", sep = "")))
}